server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java

```diff
@@ -341,24 +359,34 @@ public void onFailure(Exception e) {
         });
     }
 
-    public void deleteStaleTranslogMetadataFilesAsync() {
-        transferService.listAllAsync(ThreadPool.Names.REMOTE_PURGE, remoteMetadataTransferPath, new ActionListener<>() {
-            @Override
-            public void onResponse(Set<String> metadataFiles) {
-                List<String> sortedMetadataFiles = metadataFiles.stream().sorted(METADATA_FILENAME_COMPARATOR).collect(Collectors.toList());
-                if (sortedMetadataFiles.size() <= 1) {
-                    logger.trace("Remote Metadata file count is {}, so skipping deletion", sortedMetadataFiles.size());
-                    return;
-                }
-                List<String> metadataFilesToDelete = sortedMetadataFiles.subList(0, sortedMetadataFiles.size() - 1);
-                deleteMetadataFilesAsync(metadataFilesToDelete);
-            }
+    public void deleteStaleTranslogMetadataFilesAsync(Runnable onCompletion) {
+        try {
+            transferService.listAllInSortedOrderAsync(
+                ThreadPool.Names.REMOTE_PURGE,
+                remoteMetadataTransferPath,
+                Integer.MAX_VALUE,
+                new ActionListener<>() {
+                    @Override
+                    public void onResponse(List<BlobMetadata> blobMetadata) {
+                        List<String> sortedMetadataFiles = blobMetadata.stream().map(BlobMetadata::name).collect(Collectors.toList());
+                        if (sortedMetadataFiles.size() <= 1) {
```

Comment:
this condition should not be true because there are generations to be deleted, and if that is true then this means that there will be atleast 1 metadata file to delete, right? Not expecting you change this, but lets think once if this is needed.

Reply:
In case the tlog files get deleted , but md file deletion fails, this can still happen in future deletes.